### PR TITLE
refactor(monitor): update ovsdb event

### PR DESCRIPTION
**Pull request by echkenluo**
Signed-off-by: echken <chengcheng.luo@smartx.com>

### CODE REVIEW

File: pkg/monitor/ovsdb.go [hunk 1]
```diff
@@ -160,411 +162,431 @@ func (monitor *OVSDBMonitor) startOvsdbMonitor() error {
...
-			if _, ok := iface.Fields["external_ids"].(ovsdb.OvsMap).GoMap[LocalEndpointIdentity]; ok {
-				return true, monitor.ovsdbCache["Interface"][i]
+func (monitor *OVSDBMonitor) getPortBridgeName(portUUID string) string {
+	var bridgeName string
+	for _, bridge := range monitor.ovsdbCache["Bridge"] {
+		portUUIDs := listUUID(bridge.Fields["ports"])
+		for _, uuid := range portUUIDs {
+			if uuid.GoUuid == portUUID {
+				bridgeName = bridge.Fields["name"].(string)
+				break
```
Comment (qingwusunny) [inline pkg/monitor/ovsdb.go:579]: break can't jump double loop, maybe you can return directly

return bridgeName

File: pkg/monitor/ovsdb.go [hunk 2]
```diff
@@ -160,411 +162,431 @@ func (monitor *OVSDBMonitor) startOvsdbMonitor() error {
...
-}
-
-func (monitor *OVSDBMonitor) filterEndpoint(rowupdate ovsdb.RowUpdate) (*datapath.Endpoint, *datapath.Endpoint) {
-	newExternalIds := rowupdate.New.Fields["external_ids"].(ovsdb.OvsMap).GoMap
-	_, ok := newExternalIds[LocalEndpointIdentity]
-	if !ok {
-		if rowupdate.New.Fields["mac_in_use"] == "" {
+	// access to trunk
+	if newTag == nil && oldTag != nil && len(newTrunk) != 0 && len(oldTrunk) == 0 {
+		oldEndpoint, ok = monitor.endpointMap[ifaceUUID]
```
[the remark this answers is not in the document]
Comment (echkenluo) [reply pkg/monitor/ovsdb.go:204]: ok

File: pkg/monitor/ovsdb.go [hunk 3]
```diff
@@ -160,411 +162,431 @@ func (monitor *OVSDBMonitor) startOvsdbMonitor() error {
...
+
+	oldTrunks, ok := rowupdate.Old.Fields["trunks"].(ovsdb.OvsSet)
+	if ok {
+		trunkSet := oldTrunks.GoSet
+		for _, item := range trunkSet {
+			oldTrunk = append(oldTrunk, item.(float64))
+		}
+	}
+
+	if reflect.DeepEqual(newTrunk, oldTrunk) {
```
[the remark this answers is not in the document]
Comment (echkenluo) [reply pkg/monitor/ovsdb.go:298]: Trunk list data is in ascending order。

File: pkg/monitor/ovsdb.go (hunk 1 above)
Comment (echkenluo) [reply pkg/monitor/ovsdb.go:579]: ok

**Review by qingwusunny — APPROVED** [review]

File: pkg/monitor/ovsdb.go [hunk 4]
```diff
@@ -160,411 +162,424 @@ func (monitor *OVSDBMonitor) startOvsdbMonitor() error {
...
-		}
+func (monitor *OVSDBMonitor) filterPortVlanTagUpdate(rowupdate ovsdb.RowUpdate, ifaceUUID string) (*datapath.Endpoint, *datapath.Endpoint) {
+	var newEndpoint, oldEndpoint *datapath.Endpoint
+	var oldTag, newTag float64
+	if rowupdate.New.Fields["tag"] == nil {
+		return nil, nil
+	}
+	if rowupdate.Old.Fields["tag"] == nil {
+		return nil, nil
+	}
```
Comment (KeyongSun) [inline pkg/monitor/ovsdb.go:243]: combine them together with "||"

File: pkg/monitor/ovsdb.go [hunk 5]
```diff
@@ -160,411 +162,424 @@ func (monitor *OVSDBMonitor) startOvsdbMonitor() error {
...
-	defer monitor.cacheLock.RUnlock()
-	for _, port := range monitor.ovsdbCache["Port"] {
-		if port.Fields["name"].(string) == interfaceName {
-			return true
+		// Is this case exsit
+		if monitor.isEndpointReady(oldEndpoint) && monitor.isEndpointReady(newEndpoint) {
+			monitor.ovsdbEventHandler.UpdateLocalEndpoint(newEndpoint, oldEndpoint)
+		}
+		if monitor.isEndpointReady(newEndpoint) && !monitor.isEndpointReady(oldEndpoint) {
+			monitor.ovsdbEventHandler.AddLocalEndpoint(newEndpoint)
```
Comment (KeyongSun) [inline pkg/monitor/ovsdb.go:426]: In L415 ~ 419 if newEndpoint not assign a value, the default value will be "nil", will check the newEndpint cause error?

File: pkg/monitor/ovsdb.go [hunk 6]
```diff
@@ -160,411 +162,424 @@ func (monitor *OVSDBMonitor) startOvsdbMonitor() error {
...
+	return bridgeName
 }
 
-func (monitor *OVSDBMonitor) processEndpointDel(rowupdate ovsdb.RowUpdate) {
-	deletedEndpoints := monitor.filterEndpointDeleted(rowupdate)
-	if deletedEndpoints != nil {
-		go monitor.ovsdbEventHandler.DeleteLocalEndpoint(deletedEndpoints)
-	}
+func (monitor *OVSDBMonitor) isEndpointReady(endpoint *datapath.Endpoint) bool {
+	return endpoint.BridgeName != "" && endpoint.InterfaceUUID != "" &&
```
Comment (KeyongSun) [inline pkg/monitor/ovsdb.go:581]: Do you need  to check the endpoint is nil or not?

File: pkg/monitor/ovsdb.go (hunk 5 above)
Comment (echkenluo) [reply pkg/monitor/ovsdb.go:426]: yes, we should init oldendpoint in L410, and init newEndpoint in L416,    I forget it.

actually， this is an exception case which only happens when manually operating OVSDB, it won't normally update the interface uuid in the same port.

File: pkg/monitor/ovsdb.go (hunk 6 above)
Comment (echkenluo) [reply pkg/monitor/ovsdb.go:581]: We guarantee that the value passed in will not be nil at where the function is called.

File: pkg/monitor/ovsdb.go (hunk 4 above)
Comment (echkenluo) [reply pkg/monitor/ovsdb.go:243]: ok

**Review by qingwusunny — APPROVED** [review]

**Review by KeyongSun — APPROVED** [review]

**Review by zwtop — APPROVED** [review]

File: pkg/monitor/ovsdb.go [hunk 7]
```diff
@@ -160,411 +162,423 @@ func (monitor *OVSDBMonitor) startOvsdbMonitor() error {
...
+func (monitor *OVSDBMonitor) filterPortVlanTrunkUpdate(rowupdate ovsdb.RowUpdate, ifaceUUID string) (*datapath.Endpoint, *datapath.Endpoint) {
+	var newEndpoint, oldEndpoint *datapath.Endpoint
+	var oldTrunk, newTrunk []float64
+
+	if _, ok := monitor.endpointMap[ifaceUUID]; !ok {
 		return nil, nil
 	}
-	oldOfPort, ok := rowupdate.Old.Fields["ofport"].(float64)
-	if !ok {
+	if rowupdate.New.Fields["trunks"] == nil || rowupdate.Old.Fields["trunks"] == nil {
```
Comment (wu-cl) [inline pkg/monitor/ovsdb.go:272]: what about old is nil and new is not nil, is there another "new" logic

**Pull request merged**